dm/master/etcd_test.go

```diff
@@ -0,0 +1,187 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package master
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/pingcap/check"
+	"github.com/pingcap/pd/pkg/tempurl"
+	"go.etcd.io/etcd/embed"
+
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+var _ = check.Suite(&testEtcdSuite{})
+
+type testEtcdSuite struct {
+}
+
+func (t *testEtcdSuite) TestPrepareJoinEtcd(c *check.C) {
+	cfgCluster := NewConfig() // used to start an etcd cluster
+	cfgCluster.Name = "dm-master-1"
+	cfgCluster.DataDir = c.MkDir()
+	cfgCluster.MasterAddr = tempurl.Alloc()[len("http://"):]
+	cfgCluster.PeerUrls = tempurl.Alloc()
+	c.Assert(cfgCluster.adjust(), check.IsNil)
+
+	cfgBefore := t.cloneConfig(cfgCluster) // before `prepareJoinEtcd` applied
+	cfgBefore.DataDir = c.MkDir()          // overwrite some config items
+	cfgBefore.MasterAddr = tempurl.Alloc()[len("http://"):]
+	cfgBefore.PeerUrls = tempurl.Alloc()
+	cfgBefore.AdvertisePeerUrls = cfgBefore.PeerUrls
+	c.Assert(cfgBefore.adjust(), check.IsNil)
+
+	cfgAfter := t.cloneConfig(cfgBefore) // after `prepareJoinEtcd applied
+
+	joinCluster := cfgCluster.PeerUrls
+	joinFP := filepath.Join(cfgBefore.DataDir, "join")
+	memberDP := filepath.Join(cfgBefore.DataDir, "member")
+
+	// not set `join`, do nothing
+	c.Assert(prepareJoinEtcd(cfgAfter), check.IsNil)
+	c.Assert(cfgAfter, check.DeepEquals, cfgBefore)
+
+	// try to join self
+	cfgAfter.Join = cfgAfter.AdvertisePeerUrls
+	err := prepareJoinEtcd(cfgAfter)
+	c.Assert(terror.ErrMasterJoinEmbedEtcdFail.Equal(err), check.IsTrue)
+	c.Assert(err, check.ErrorMatches, ".*fail to join embed etcd: join self.*is forbidden.*")
+
+	// update `join` to a valid item
+	cfgBefore.Join = joinCluster
+
+	// join with persistent data
+	c.Assert(ioutil.WriteFile(joinFP, []byte(joinCluster), privateDirMode), check.IsNil)
+	cfgAfter = t.cloneConfig(cfgBefore)
```

Comment:
if there are not a persistent data, `c.Assert(cfgAfter.InitialCluster, check.Equals, joinCluster)` also pass (because of L170 of etcd.go)

Reply:
🤔 Did you mean
```
cfg.InitialCluster = strings.Join(ms, ",")
```

if so, it's verified by `// join with existing cluster` from L122 to L129 of etcd_test.go